### ISSUE TITLE
error: Incompatible return value type (got "Union[Dict[Any, Any], List[Dict[Any, Any]]]", expected "Dict[Any, Any]")

### ISSUE THREAD

**Issue by janwadolowski** [python/mypy#11692]
I think it's a bug or am I missing something?

```
from typing import List, Union, Type, Dict

from cassandra.cqlengine.models import Model
from marshmallow import Schema


def serialize_object_to_json(
    schema: Type[Schema], db_object: Union[List[Model], Model]
) -> Union[Dict, List[Dict]]:
    if isinstance(db_object, List):
        return schema().dump(obj=db_object, many=True)
    elif isinstance(db_object, Model):
        return schema().dump(obj=db_object, many=False)
    else:
        raise TypeError
```

**Comment by janwadolowski** [comment]
Also the scenario:


```
from typing import List, Union, Type, Dict
from typing import overload

from cassandra.cqlengine.models import Model
from marshmallow import Schema


@overload
def serialize_object_to_json(
    schema: Type[Schema], db_object: List[Model]
) -> List[Dict]:
    ...


@overload
def serialize_object_to_json(schema: Type[Schema], db_object: Model) -> Dict:
    ...


def serialize_object_to_json(
    schema: Type[Schema], db_object: Union[List[Model], Model]
) -> Union[Dict, List[Dict]]:
    if isinstance(db_object, List):
        return schema().dump(obj=db_object, many=True)
    elif isinstance(db_object, Model):
        return schema().dump(obj=db_object, many=False)
    else:
        raise TypeError
```

resulting in an error: Overloaded function signatures 1 and 2 overlap with incompatible return types

**Comment by sobolevn** [comment]
Can you please post a simplier reproduction? Right now it is quite hard to follow what is going on, since your code depends on some 3rd party APIs.

**Comment by janwadolowski** [comment]
Here you are:
from typing import Union, List


```
def single_or_multiple(text: str) -> Union[List[str], str]:
    output = text.split()
    return output if len(output) > 1 else output[0]


def only_accept_str() -> None:
    some_var: List[str] = single_or_multiple("Mypy why are you doing this?")
    other_var: str = single_or_multiple("help")
```


test.py:10: error: Incompatible types in assignment (expression has type "Union[List[str], str]", variable has type "List[str]")
test.py:11: error: Incompatible types in assignment (expression has type "Union[List[str], str]", variable has type "str")

**Comment by janwadolowski** [comment]
Also occurs when you make output explicitly dependable on a boolean flag like many: bool.

**Comment by sobolevn** [comment]
@janwadolowski this is not a bug, this is how typing should be. Let me exaplain 🙂 

This problem is known as "union return type": https://github.com/python/mypy/issues/1693
So, when you write `-> Union[List[str], str]:` - you say: this function can return either `List[str]` or `str` and we don't know when which type is returned.

But, when you write `some_var: List[str] = single_or_multiple("Mypy why are you doing this?")` - you say "I am sure that `some_var` is `List[str]`". But, the right hand side still has `Union[List[str], str]` type. And we have a mismatch.

To fix this, there are two ways:
1. `if isinstance(some_var, str):` to work with `str` values separately and with `List[str]` separately
2. Use `cast()`, like so: `some_var = cast(List[str], single_or_multiple("Mypy why are you doing this?"))`

I hope this help! 👍

**Comment by janwadolowski** [comment]
Thanks that actually helped me wrap my head around it. Of the two I definitelly prefer and will implement if isinstance as it actually introduce some "type security" logic, whereas cast() can only shut up mypy errors. Thank you again, it's clear for me now.
